rust/src/mysql/parser.rs

```diff
@@ -0,0 +1,3350 @@
+            let (i, value) = map(take(value_len as usize), |s: &[u8]| {
+                String::from_utf8_lossy(s).to_string()
+            })(i)?;
+            res.push(MysqlHandshakeResponseAttribute { key, value });
+            length -= key_len + value_len;
+            rem = i;
+        }
+
+        Ok((rem, res))
+    }
+}
+
+pub fn parse_auth_response(i: &[u8]) -> IResult<&[u8], MysqlResponse> {
+    let (i, header) = parse_packet_header(i)?;
+    let (i, status) = be_u8(i)?;
+    match status {
+        0x00 => {
+            let (i, response) = parse_response_ok(i, header.pkt_len - 1)?;
+            Ok((i, response))
+        }
+        // AuthMoreData
+        0x01 => {
+            let (i, data) = be_u8(i)?;
+            let (i, _) = cond(header.pkt_len - 2 > 0, take(header.pkt_len - 2))(i)?;
+            Ok((
+                i,
+                MysqlResponse {
+                    item: MysqlResponsePacket::AuthMoreData { data },
+                },
+            ))
+        }
+        0xEF => {
+            let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+            Ok((
+                i,
+                MysqlResponse {
+                    item: MysqlResponsePacket::EOF,
+                },
+            ))
+        }
+        _ => {
+            let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+            Ok((
+                i,
+                MysqlResponse {
+                    item: MysqlResponsePacket::Unknown,
+                },
+            ))
+        }
+    }
+}
+
+pub fn parse_auth_switch_request(i: &[u8]) -> IResult<&[u8], MysqlAuthSwtichRequest> {
+    let (i, header) = parse_packet_header(i)?;
+    let plugin_length = i.len();
+    let (i, plugin_name) = map(take_till(|ch| ch == 0x00), |s: &[u8]| {
+        String::from_utf8_lossy(s).to_string()
+    })(i)?;
+    let plugin_length = plugin_length - i.len();
+    let (i, plugin_data) = map(
+        cond(
+            header.pkt_len - (plugin_length as u32) > 0,
+            take(header.pkt_len - plugin_length as u32),
+        ),
+        |ch: Option<&[u8]>| {
+            if let Some(ch) = ch {
+                String::from_utf8_lossy(ch).to_string()
+            } else {
+                String::new()
+            }
+        },
+    )(i)?;
+
+    Ok((
+        i,
+        MysqlAuthSwtichRequest {
+            plugin_name,
+            plugin_data,
+        },
+    ))
+}
+
+pub fn parse_local_file_data_content(i: &[u8]) -> IResult<&[u8], u32> {
+    let (i, header) = parse_packet_header(i)?;
+    let (i, _) = take(header.pkt_len)(i)?;
+    Ok((i, header.pkt_len))
+}
+
+pub fn parse_request(
+    i: &[u8], params: Option<u16>, param_types: Option<Vec<MysqlColumnDefinition>>,
+    stmt_long_datas: Option<Vec<StmtLongData>>, client_flags: u32,
+) -> IResult<&[u8], MysqlRequest> {
+    let (i, header) = parse_packet_header(i)?;
+    let (i, command_code) = be_u8(i)?;
+    match command_code {
```

Comment:
This can be rewritten as:
```
let (i, req) = match command_code {
    0x01 => (i, MysqlRequest {
                command_code,
                command: MysqlCommand::Quit,
    })
    ...
}
Ok((i, req))
```

Reply:
Fixed, please check again.